### PR TITLE
Wire up credential helper to command-line flag(s)

### PR DESCRIPTION
**Pull request by Yannic**
RELNOTES: Add support for fetching RPC credentials from credential helper.

Progress on https://github.com/bazelbuild/bazel/issues/15856

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java [hunk 1]
```diff
@@ -139,6 +139,36 @@ public class AuthAndTLSOptions extends OptionsBase {
...
+
+  @Option(
+      name = "credential_helper_cache_timeout",
+      defaultValue = "15m",
+      converter = DurationConverter.class,
+      documentationCategory = OptionDocumentationCategory.UNCATEGORIZED,
+      effectTags = {OptionEffectTag.UNKNOWN},
+      help = "TODO")
+  public Duration credentialHelperCacheTimeout;
+
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:171]: Don't forget to mark these flags `experimental` as well.

File: src/main/java/com/google/devtools/build/lib/authandtls/GoogleAuthUtils.java [hunk 2]
```diff
@@ -234,24 +235,37 @@ public static CallCredentialsProvider newCallCredentialsProvider(@Nullable Crede
    */
   @Nullable
   public static Credentials newCredentials(
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/GoogleAuthUtils.java:237]: Please update the comment.

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  public CredentialHelperCredentials(
+      CredentialHelperProvider credentialHelperProvider,
+      CredentialHelperEnvironment credentialHelperEnvironment,
+      Optional<Credentials> fallbackCredentials,
+      Duration cacheTimeout) {
+    Preconditions.checkNotNull(credentialHelperProvider);
+    Preconditions.checkNotNull(credentialHelperEnvironment);
+    this.fallbackCredentials = Preconditions.checkNotNull(fallbackCredentials);
+    Preconditions.checkNotNull(cacheTimeout);
+    Preconditions.checkArgument(!cacheTimeout.isNegative() && !cacheTimeout.isZero(), "Cache timeout must be greater than 0");
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java:34]: Just to confirm, does the flag parsing logic already reject negative values? (i.e., is this checking for a programmer error or a user error?)

**Review by tjgq — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Implementation of {@link Credentials} which fetches credentials by invoking a {@code credential
+ * helper} as subprocess.
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java:18]: `..., falling back to another {@link Credentials} if no suitable helper exists.`

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java (hunk 3 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java:34]: I think so: https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/common/options/DurationConverterTest.java;l=53;drc=ddcbe585e19a62a184651ca76384b468591a3428

File: src/main/java/com/google/devtools/build/lib/authandtls/GoogleAuthUtils.java (hunk 2 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/GoogleAuthUtils.java:237]: done

File: src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java (hunk 4 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/credentialhelper/CredentialHelperCredentials.java:18]: done

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java (hunk 1 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:171]: done

**Comment by Yannic** [comment]
PTAL

**Comment by Yannic** [comment]
@bazel-io flag

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java [hunk 5]
```diff
@@ -139,6 +139,41 @@ public class AuthAndTLSOptions extends OptionsBase {
...
+      defaultValue = "null",
+      allowMultiple = true,
+      converter = UnresolvedScopedCredentialHelperConverter.class,
+      documentationCategory = OptionDocumentationCategory.UNCATEGORIZED,
+      effectTags = {OptionEffectTag.UNKNOWN},
+      help =
+          "Configures Credential Helpers to use for retrieving credentials for the provided scope "
+              + "(domain).\n\n"
+              + "Credentials from Credential Helpers take precedence over credentials from "
+              + "<code>--google_default_credentials</code> or <code>.netrc</code>.\n\n"
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:153]: Also `--google_credentials`.

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java [hunk 6]
```diff
@@ -139,6 +139,41 @@ public class AuthAndTLSOptions extends OptionsBase {
...
+      documentationCategory = OptionDocumentationCategory.UNCATEGORIZED,
+      effectTags = {OptionEffectTag.UNKNOWN},
+      help =
+          "Configures the timeout for the Credential Helper.\n\n"
+              + "Credential Helpers failing to respond within this timeout will fail the invocation.")
+  public Duration credentialHelperTimeout;
+
+  @Option(
+      name = "experimental_credential_helper_cache_duration",
+      defaultValue = "15m",
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:170]: Do you mind going with 30 minutes instead?

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java [hunk 7]
```diff
@@ -139,6 +139,41 @@ public class AuthAndTLSOptions extends OptionsBase {
               + "pings are disabled, then this setting is ignored.")
   public Duration grpcKeepaliveTimeout;
 
+  @Option(
+      name = "experimental_experimental_credential_helper",
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:143]: Remove the extra `experimental`.

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java [hunk 8]
```diff
@@ -139,6 +139,41 @@ public class AuthAndTLSOptions extends OptionsBase {
...
+      help =
+          "Configures Credential Helpers to use for retrieving credentials for the provided scope "
+              + "(domain).\n\n"
+              + "Credentials from Credential Helpers take precedence over credentials from "
+              + "<code>--google_default_credentials</code> or <code>.netrc</code>.\n\n"
+              + "See https://github.com/bazelbuild/proposals/blob/main/designs/2022-06-07-bazel-credential-helpers.md for details.")
+  public List<UnresolvedScopedCredentialHelper> credentialHelpers;
+
+  @Option(
+      name = "credential_helper_timeout",
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:158]: Add `experimental` here.

**Review by tjgq — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java (hunk 6 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:170]: Sure

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java (hunk 7 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java:143]: Of course I copy&pasted to the wrong flag 🤦‍♀️

File: src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java (hunk 5 above)
Comment (Yannic) [reply src/main/java/com/google/devtools/build/lib/authandtls/AuthAndTLSOptions.java]: done

**Review by tjgq — APPROVED** [review]

**Pull request closed without merging**